conn.go

```diff
@@ -349,72 +340,60 @@ func (c *Conn) start() error {
 	// this is because our peer can tell us the max channels they support.
 	c.channels = bitmap.New(uint32(c.channelMax))
 
-	// start multiplexor and writer
-	go c.mux()
 	go c.connWriter()
+	go c.connReader()
 
 	return nil
 }
 
 // Close closes the connection.
 func (c *Conn) Close() error {
-	c.closeMuxOnce.Do(func() { close(c.closeMux) })
-	err := c.err()
+	c.closeOnce.Do(func() { c.close() })
+
 	var connErr *ConnError
-	if errors.As(err, &connErr) && connErr.RemoteErr == nil && connErr.inner == nil {
+	if errors.As(c.doneErr, &connErr) && connErr.RemoteErr == nil && connErr.inner == nil {
 		// an empty ConnectionError means the connection was closed by the caller
-		// or as requested by the peer and no error was provided in the close frame.
 		return nil
 	}
-	return err
+
+	// there was an error during shut-down or connReader/connWriter
+	// experienced a terminal error
+	return c.doneErr
 }
 
-// close should only be called by conn.mux.
+// close is called once, either from Close() or when connReader/connWriter exits
 func (c *Conn) close() {
-	close(c.done) // notify goroutines and blocked functions to exit
+	defer close(c.done)
+
+	close(c.rxtxExit)
 
 	// wait for writing to stop, allows it to send the final close frame
 	<-c.txDone
 
-	// reading from connErr in mux can race with closeMux, causing
-	// a pending conn read/write error to be lost.  now that the
-	// mux has exited, drain any pending error.
-	select {
-	case err := <-c.connErr:
-		c.doneErr = err
-	default:
-		// no pending read/write error
-	}
-
-	err := c.net.Close()
-	switch {
-	// conn.err already set
-	// TODO: err info is lost, log it?
-	case c.doneErr != nil:
-
-	// conn.err not set and c.net.Close() returned a non-nil error
-	case err != nil:
-		c.doneErr = err
-
-	// no errors
-	default:
-	}
+	closeErr := c.net.Close()
 
 	// check rxDone after closing net, otherwise may block
 	// for up to c.idleTimeout
 	<-c.rxDone
-}
 
-// Err returns the connection's error state after it's been closed.
-// Calling this on an open connection will block until the connection is closed.
-func (c *Conn) err() error {
-	c.doneErrMu.Lock()
-	defer c.doneErrMu.Unlock()
-	var amqpErr *Error
-	if errors.As(c.doneErr, &amqpErr) {
-		return &ConnError{RemoteErr: amqpErr}
+	if errors.Is(c.rxErr, net.ErrClosed) {
+		// this is the expected error when the connection is closed, swallow it
+		c.rxErr = nil
+	}
+
+	if c.txErr == nil && c.rxErr == nil && closeErr == nil {
+		// if there are no errors, it means user initiated close() and we shut down cleanly
+		c.doneErr = &ConnError{}
+	} else if amqpErr, ok := c.rxErr.(*Error); ok {
+		// we experienced a peer-initiated close that contained an Error.  return it
+		c.doneErr = &ConnError{RemoteErr: amqpErr}
+	} else if c.txErr != nil {
```

Comment:
I couldn't think of a reason but is there any significance to which err we check first? (txErr vs rxErr)

Reply:
Yes, but only insofar as what error message is exposed to the caller via the `ConnError` in the event that `connReader()` and `connWriter()` experienced terminal errors.

I debated about concatenating them both together but decided against it as I felt the added complexity (which admittingly is minor) wasn't worth it (i.e. how actionable would the error message really be).  I could be convinced to do this though if you think there's a compelling reason.